File: 162.find-peak-element.js
/*
 * @lc app=leetcode id=162 lang=javascript
 *
 * [162] Find Peak Element
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var findPeakElement = function(nums) {
  // O(log(n)) time
  // O(k) space
  let low = 0
  let high = nums.length - 1
  while (low < high) {
    let midPoint = Math.floor((high + low) / 2)
    let midPoint2 = midPoint+1
    if (nums[midPoint] < nums[midPoint2]) {
      low = midPoint2
    } else {
      high = midPoint
    }
  }
  return low
};
// @lc code=end

File: cpp/162.find-peak-element.cpp
/*
 * @lc app=leetcode id=162 lang=cpp
 *
 * [162] Find Peak Element
 */

// @lc code=start
class Solution
{
public:
    int findPeakElement(vector<int> &nums)
    {
        // revisit when wiser, this is wild
        // O(log(n)) time O(k) space
        int indexL = 0;
        int indexH = nums.size() - 1;
        while (indexL < indexH)
        {
            int midPoint = (indexH + indexL) / 2;
            int midPoint2 = midPoint + 1;
            if (nums[midPoint] < nums[midPoint2])
            {
                indexL = midPoint2;
            }
            else
            {
                indexH = midPoint;
            }
        }
        return indexL;
    }
};
// @lc code=end
